### PR TITLE
Document the live migration process

### PR DESCRIPTION
**Pull request by vladikr**
Signed-off-by: Vladik Romanovsky <vromanso@redhat.com>

### CODE REVIEW

File: administration/live-migration.adoc [hunk 1]
```diff
@@ -0,0 +1,138 @@
+# Live Migration
+
+Live migration is a process during which a running Virtual Machine Instance
+moves to another compute node while the guest workload continues to run and
+remain accessable.
+
+## Initiate live migration
```
Comment (fabiand) [inline administration/live-migration.adoc]: Does a human operator need to open some feature gate in order to use LM?

File: administration/live-migration.adoc [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+`vmi-fedora`
+
+```yaml
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachineInstanceMigration
+metadata:
+  name: migration-job
+spec:
+  vmiName: vmi-fedora
+configuration:
```
Comment (fabiand) [inline administration/live-migration.adoc]: Is this a top-level item? `VMIM.configuration`?

File: administration/live-migration.adoc [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  completionTimeoutPerGiB: 800
+  progressTimeout: 150
+```
+
+An optional configuration can be provided as part of the `VirtualMachineInstanceMigration` object
+
+### configuration
```
Comment (fabiand) [inline administration/live-migration.adoc]: **C**onfiguration

File: administration/live-migration.adoc [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+Depending on the type, live migration process will copy virtual machine memory
+pages and disk blocks To the destination.
```
Comment (fabiand) [inline administration/live-migration.adoc]: **to**

File: administration/live-migration.adoc [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+configuration:
+  completionTimeoutPerGiB: 800
+  progressTimeout: 150
+```
+
+An optional configuration can be provided as part of the `VirtualMachineInstanceMigration` object
+
+### configuration
+
+Depending on the type, live migration process will copy virtual machine memory
```
Comment (fabiand) [inline administration/live-migration.adoc]: **the** live migration process …

File: administration/live-migration.adoc [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+In this case, live migration will be aborted if it is running for a long
+perioud of time. The timeout is calculated base on the side of the VMI, it's
+memory and the ephemeral disks that are needed to be copied. The configurable
+parameter `completionTimeoutPerGiB`, which deafults to 800s is the time for GiB
+of data to wait for the migration to be completed before aborting it.
+A VMI with 8Gib of memory will time out after 6400 seconds.
+
+#### Progress timeout
+
+Live migration will also be aborted when it will be noticed that memory doesn't
```
Comment (fabiand) [inline administration/live-migration.adoc]: … that **copying** memory …

File: administration/live-migration.adoc [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+The time to wait for live migration to make progress in transferring data is
+configurable by `progressTimeout` parameter, which defaults to 150s
+
+## Migration Status Reporting
+
+### Condition and migration method
+
+When starting a virtual machine instance, it has also been calculated whether
+the machine is live migratable. The result is being stored in the VMI
+`Conditions`. The calculation can be based on multiple parameters of the VMI,
```
Comment (fabiand) [inline administration/live-migration.adoc]: Maybe use the jsonpath `VMI.status.conditions`?

File: administration/live-migration.adoc [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+Status:
+  Conditions:
+    Status:                True
+    Type:                  LiveMigratable
+  Migration Method:  BlockMigration
+```
+
+### Migration Status
+
+The migration progress status is being reported in the VMI `Status`.
```
Comment (fabiand) [inline administration/live-migration.adoc]: Maybe jsonpath again: `VMI.status`

File: administration/live-migration.adoc [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+Most importantly, it indicates whether the migration has been `Completed` and
```
Comment (fabiand) [inline administration/live-migration.adoc]: … **or** if it failed …

File: administration/live-migration.adoc [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+Migration State:
+    Completed:        true
+    End Timestamp:    2019-03-29T03:37:52Z
+    Migration Config:
+      Completion Timeout Per GiB:  800
+      Progress Timeout:             150
+    Migration UID:                  c64d4898-51d3-11e9-b370-525500d15501
+    Source Node:                    node02
+    Start Timestamp:                2019-03-29T04:02:47Z
+    Target Direct Migration Node Ports:
```
Comment (fabiand) [inline administration/live-migration.adoc]: Do we need these informations in the API?

File: administration/live-migration.adoc [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+      35001:                      0
+      41068:                      49152
+      38284:                      49153
+    Target Node:                  node01
+    Target Node Address:          10.128.0.46
```
Comment (fabiand) [inline administration/live-migration.adoc]: Do we need the IP as well?

If: Should it be `TargetNodeIPAddress`

File: administration/live-migration.adoc [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    Target Node Domain Detected:  true
```
Comment (fabiand) [inline administration/live-migration.adoc]: Do we need to expose this information?

File: administration/live-migration.adoc [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    Target Pod:                   virt-launcher-testvmimcbjgw6zrzcmp8wpddvztvzm7x2k6cjbdgktwv8tkq
```
Comment (fabiand) [inline administration/live-migration.adoc]: Do we need to expose this information?

File: administration/live-migration.adoc (hunk 10 above)
Comment (vladikr) [reply administration/live-migration.adoc]: That's in the vmi.migrationState. Well, yes, this is needed at this point, as this is how the source knows which ports to use on the destination.

File: administration/live-migration.adoc (hunk 11 above)
Comment (vladikr) [reply administration/live-migration.adoc]: Yes, also needed to for the migration proxy.

Re: `TargetNodeIPAddress`, can easily be changed, but not in the doc.

File: administration/live-migration.adoc (hunk 13 above)
Comment (vladikr) [reply administration/live-migration.adoc]: Yes, it's needed for the source pod to resolve the destination pod.

File: administration/live-migration.adoc (hunk 1 above)
Comment (vladikr) [reply administration/live-migration.adoc]: Yep, I've missed it. Added.

File: administration/live-migration.adoc (hunk 2 above)
Comment (vladikr) [reply administration/live-migration.adoc]: fixed.

File: administration/live-migration.adoc (hunk 5 above)
Comment (vladikr) [reply administration/live-migration.adoc]: Done

File: administration/live-migration.adoc (hunk 6 above)
Comment (vladikr) [reply administration/live-migration.adoc]: Done

File: administration/live-migration.adoc (hunk 7 above)
Comment (vladikr) [reply administration/live-migration.adoc]: done

File: administration/live-migration.adoc (hunk 8 above)
Comment (vladikr) [reply administration/live-migration.adoc]: done

File: administration/live-migration.adoc (hunk 9 above)
Comment (vladikr) [reply administration/live-migration.adoc]: done

**Comment by vladikr** [comment]
@fabiand I've removed the configuration part.

File: administration/live-migration.adoc [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+`vmi-fedora`
+
+```yaml
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachineInstanceMigration
+metadata:
+  name: migration-job
+spec:
+  vmiName: vmi-fedora
+  configuration:
```
Comment (rmohr) [inline administration/live-migration.adoc]: Is this still supposed to be there?

**Comment by fabiand** [comment]
yey.

**Pull request merged**